### ISSUE TITLE
Heredoc containing a closing PHP tag breaks syntax highlighting

### ISSUE THREAD

**Issue by jaikdean** [atom/language-php#10]
I have a file containing the following:

``` php
        $this->xml = <<<EOD
<?xml version="1.0"?>
<methodCall>
<methodName>{$this->method}</methodName>
<params>

EOD;
```

The syntax formatting is picking up the `?>` within the heredoc as a closing PHP tag, and as such the syntax highlighting is broken from this point on in the file.

![screen shot 2014-03-03 at 10 58 40](https://f.cloud.github.com/assets/348408/2308690/d8630ad8-a2c2-11e3-9fdb-bd1544bb38bc.png)

**Comment by nicksayre** [comment]
Syntax highlighting also breaks with inclusion of a single-quote in a heredoc.

![screen shot 2015-09-16 at 10 08 12 am](https://cloud.githubusercontent.com/assets/1807967/9908774/c575671a-5c5a-11e5-981d-7a0ffd4d4b58.png)

**Comment by TheJosh** [comment]
Likewise for single quotes (`'`) or double quotes (`"`) within a heredoc

**Comment by zr9** [comment]
There is 2 problems in atom/core probably.
Because grammar ignore "repository" and "injection" rules.

Good news so i'm can fix it by rewriting without that rules, here is simple trial example
![atom_issue_3](https://cloud.githubusercontent.com/assets/13107923/12418160/6fce551c-bebe-11e5-8d6e-070b61e692cb.png)

But i'm need someone to help me write bug about grammar(En is not my native, so i'm sometimes unclear as hell).

And as well i'm need someone who can explain tricky parts in itself heredoc syntax,
like scope "string.unquoted.heredoc.php" now contain in rule **double quote** 
and scope "string.unquoted.heredoc.nowdoc.php" now contain **single quote**
so i'm need to know diff and probably some examples of different languages(Js,SQL,Json,Regex,CSS) heredoc to test

**Comment by Ingramz** [comment]
@zr9 we've already concluded that it would be worthwhile fixing the root cause of this problem in first-mate package ([see related discussion](https://github.com/atom/language-php/pull/77)), so that all other bugs would fix on their own. The current solution is rather elegant (although, not working).

But it's still nice seeing someone giving their go at it.

**Comment by zr9** [comment]
@Ingramz Oh i'm not read references, so core bug already submitted. Then np :) But current solution is seems like working, so it can be fine as temporally fix until atom fixed "repository"

**Comment by Ingramz** [comment]
@zr9 sure, that's something people have suggested before.

You can compare your solution to the one in the pull request I submitted above. The changes got undone with later commits, but if you keep track of the commits throughout the pull request, you can sort-of see what was done there.

**Comment by zr9** [comment]
@Ingramz Yup almost similar, difference in "interpolation" since injection rule is not work, so interpolation need to get moved inside "heredoc_interior" after that it get fully working in current state, because at first heredoc get covered by main rules, then interpolated by php rules inside. Look on screen i'm posted above, or i'm can attach code if you wish so. So technically in current state full fix possible, and it even will be not dirty, only dirty thing is "interpolation" repeat. And more when repository will be fixed in core, it continue to work, so not get broken.

**Comment by zr9** [comment]
@Ingramz Tested, looks totally fine, here is screenshots.

![heredoc_1](https://cloud.githubusercontent.com/assets/13107923/12420212/5e8d9994-becc-11e5-947e-f40ee50de013.png)
![heredoc_2](https://cloud.githubusercontent.com/assets/13107923/12420215/64428444-becc-11e5-9342-1a2920accb4d.png)

Except of color issues, but that is atom/core other bug(or not bug).
Scopes for brief look fine too.

**Comment by zr9** [comment]
@Ingramz So should i'm push commit to fix this one? Cos there is no seems any activity on repository fix in mate

**Comment by Ingramz** [comment]
@zr9 sure, you can do it.

**Comment by zr9** [comment]
@Ingramz Okay then once @50Wliu merge things, i'm clean file, and then update this one

**Comment by diego-vieira** [comment]
@zr9 did you sent a PR? having the same issue here.
Thanks

**Comment by zr9** [comment]
@diego-vieira Not yet, have  a lot less time now, so slowdown a much, but still have plans to sent PR, or you can do so if you wish to.

**Comment by kwladyka** [comment]
<img width="418" alt="screen shot 2016-07-24 at 23 55 18" src="https://cloud.githubusercontent.com/assets/3903726/17086741/662b3992-51fa-11e6-936f-e72805a4ae4c.png">

I can confirm problem still occur.

**Comment by 50Wliu** [comment]
Fixed by #184.
